### PR TITLE
Update memtier support

### PR DESCRIPTION
**Pull request by PatKamin**
This change is [Reviewable](https://reviewable.io/reviews/patkamin/redis/1)

### CODE REVIEW

**Comment by michalbiesek** [comment]
https://github.com/memkind/memkind/commit/4ff24e8a2d656915b1497dd9256086cf8579ff06 please update to this commit actually static could have a small performance impact on it.
I would also include the https://github.com/memkind/memkind/pull/619 but we need to merge it first

**Comment by michalbiesek** [comment]
Besides the previous comment LGTM

**Comment by PatKamin** [comment]
> [memkind/memkind@4ff24e8](https://github.com/memkind/memkind/commit/4ff24e8a2d656915b1497dd9256086cf8579ff06) please update to this commit actually static could have a small performance impact on it.
> I would also include the [memkind/memkind#619](https://github.com/memkind/memkind/pull/619) but we need to merge it first

Done.

**Comment by michalbiesek** [comment]
LGTM

File: src/zmalloc.c [hunk 1]
```diff
@@ -504,12 +506,16 @@ void zfree_dram(void *ptr) {
 
     if (ptr == NULL) return;
 #ifdef HAVE_MALLOC_SIZE
+#ifndef USE_MEMKIND_MEMTIER
     update_zmalloc_dram_stat_free(zmalloc_size(ptr));
+#endif
     free_dram(ptr);
 #else
     realptr = (char*)ptr-PREFIX_SIZE;
     oldsize = *((size_t*)realptr);
```
Comment (michalbiesek) [inline src/zmalloc.c]: You don't use oldsize variable in case of USE_MEMKIND_MEMTIER consider this in ifdef block

File: src/zmalloc.c [hunk 2]
```diff
@@ -458,8 +456,10 @@ void *zrealloc_dram(void *ptr, size_t size) {
     newptr = realloc_dram(ptr,size);
```
Comment (michalbiesek) [inline src/zmalloc.c:415]: You don't use oldsize variable in case of USE_MEMKIND_MEMTIER consider this in ifdef block

File: src/zmalloc.c [hunk 3]
```diff
@@ -383,8 +381,6 @@ static void *zrealloc_pmem(void *ptr, size_t size) {
     newptr = realloc_pmem(ptr,size);
     if (!newptr) zmalloc_oom_handler(size);
 
-    update_zmalloc_pmem_stat_free(oldsize);
```
Comment (michalbiesek) [inline src/zmalloc.c]: Why you remove this USE_MEMKIND_MEMTIER consider this in ifdef block AFAIK this block is shared with USE_MEMKIND

File: src/zmalloc.c (hunk 1 above)
Comment (PatKamin) [reply src/zmalloc.c]: Done.

File: src/zmalloc.c (hunk 2 above)
Comment (PatKamin) [reply src/zmalloc.c:415]: Done.

File: src/zmalloc.c (hunk 3 above)
Comment (PatKamin) [reply src/zmalloc.c]: I think it's not shared with a USE_MEMKIND block as there is a IF ELSE statement choosing between USE_MEMKIND_MEMTIER block and USE_MEMKIND block. Am I missing sth?

Comment (michalbiesek) [reply src/zmalloc.c]: It is ok

**Comment by michalbiesek** [comment]
LGTM

**Review by michalbiesek — COMMENTED** [review]
> Reviewed 7 of 10 files at r1, 2 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 2 unresolved discussions (waiting on @PatKamin)

___
*[src/zmalloc.c, line 91 at r3](https://reviewable.io/reviews/patkamin/redis/1#-MYiiczoDdRtSZGT5Lhh:-MYiiczoDdRtSZGT5Lhi:b-hyjqkh) ([raw file](https://github.com/patkamin/redis/blob/e902bb07d82af93c8394dc03beab11e6aab2e71b/src/zmalloc.c#L91)):*
> ```
> memkind_t dram_kind;
> memkind_t pmem_kind;
> ```
static?
___
*[src/zmalloc.c, line 95 at r3](https://reviewable.io/reviews/patkamin/redis/1#-MYij9cK8j6Nmb7ZzV50:-MYij9cK8j6Nmb7ZzV51:bfkxcun) ([raw file](https://github.com/patkamin/redis/blob/e902bb07d82af93c8394dc03beab11e6aab2e71b/src/zmalloc.c#L95)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> #define realloc_dram(ptr, size) memtier_realloc(tiered_memory, ptr, size)
> #define realloc_pmem(ptr, size) memtier_realloc(tiered_memory, ptr, size)
> #define free_dram(ptr) memtier_realloc(tiered_memory, ptr, 0)
> #define free_pmem(ptr) memtier_realloc(tiered_memory, ptr, 0)
> ```
</details>

tiered_memory - shouldn't you use here memtier_kind*API and call it by kind e.g.

Instead of:

> ```
> #define free_dram(ptr) memtier_realloc(tiered_memory, ptr, 0)
> ```

use

> ```
> #define free_dram(ptr) memtier_kind_realloc(dram_kind, ptr, 0)
> ```

**Review by PatKamin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 2 unresolved discussions (waiting on @michalbiesek)

___
*[src/zmalloc.c, line 91 at r3](https://reviewable.io/reviews/patkamin/redis/1#-MYiiczoDdRtSZGT5Lhh:-MYikVBC9VkARShxnsb3:b-896fix) ([raw file](https://github.com/patkamin/redis/blob/e902bb07d82af93c8394dc03beab11e6aab2e71b/src/zmalloc.c#L91)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> memkind_t dram_kind;
> memkind_t pmem_kind;
> ```
static?
</blockquote></details>

Done.
___
*[src/zmalloc.c, line 95 at r3](https://reviewable.io/reviews/patkamin/redis/1#-MYij9cK8j6Nmb7ZzV50:-MYiklc--m02dtsCyNKg:bo6apac) ([raw file](https://github.com/patkamin/redis/blob/e902bb07d82af93c8394dc03beab11e6aab2e71b/src/zmalloc.c#L95)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> #define realloc_dram(ptr, size) memtier_realloc(tiered_memory, ptr, size)
> #define realloc_pmem(ptr, size) memtier_realloc(tiered_memory, ptr, size)
> #define free_dram(ptr) memtier_realloc(tiered_memory, ptr, 0)
> #define free_pmem(ptr) memtier_realloc(tiered_memory, ptr, 0)
> ```
tiered_memory - shouldn't you use here memtier_kind*API and call it by kind e.g.

Instead of:

> ```
> #define free_dram(ptr) memtier_realloc(tiered_memory, ptr, 0)
> ```

use

> ```
> #define free_dram(ptr) memtier_kind_realloc(dram_kind, ptr, 0)
> ```
</blockquote></details>

Indeed.

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 3 unresolved discussions (waiting on @PatKamin)

___
*[src/zmalloc.c, line 380 at r7](https://reviewable.io/reviews/patkamin/redis/1#-MYjcDF9D-7VH7PXe8PA:-MYjcDF9D-7VH7PXe8PB:b-8hp8if) ([raw file](https://github.com/patkamin/redis/blob/4fe40ce115b3595defc483ffdd76755bc8dfc9f6/src/zmalloc.c#L380)):*
> ```
> zmalloc(size)
> ```
zmalloc_pmem
___
*[src/zmalloc.c, line 401 at r7](https://reviewable.io/reviews/patkamin/redis/1#-MYjckOp6wEn4fimTqyQ:-MYjckOp6wEn4fimTqyR:byh1cnx) ([raw file](https://github.com/patkamin/redis/blob/4fe40ce115b3595defc483ffdd76755bc8dfc9f6/src/zmalloc.c#L401)):*
> ```
>     return (size < pmem_threshold) ? zmalloc_dram(size) : zmalloc_pmem(size);
> ```

I think this is a place for something like:
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> #ifdef USE_MEMKIND_MEMTIER
>  return malloc(size);
> #else
>    return (size < pmem_threshold) ? zmalloc_dram(size) : zmalloc_pmem(size);
> #endif
> ```
</details>


But I am not sure what is the relation between pmem_threshold and MEMKIND_MEMTIER :/
___
*[src/zmalloc.c, line 444 at r7](https://reviewable.io/reviews/patkamin/redis/1#-MYjd5ByDOMy2kHkYKbF:-MYjd5ByDOMy2kHkYKbG:b-j6f7il) ([raw file](https://github.com/patkamin/redis/blob/4fe40ce115b3595defc483ffdd76755bc8dfc9f6/src/zmalloc.c#L444)):*
> ```
>     return (size < pmem_threshold) ? zcalloc_dram(size) : zcalloc_pmem(size);
> ```
> #ifdef USE_MEMKIND_MEMTIER
>  return calloc(size);
> #else
>    return (size < pmem_threshold) ?  zcalloc_dram(size) : zcalloc_pmem(size);
> #endif
> ```

**Review by PatKamin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 3 unresolved discussions (waiting on @michalbiesek)

___
*[src/zmalloc.c, line 380 at r7](https://reviewable.io/reviews/patkamin/redis/1#-MYjcDF9D-7VH7PXe8PA:-MYoa4LZ0a2rn-KP0A5V:b-896fix) ([raw file](https://github.com/patkamin/redis/blob/4fe40ce115b3595defc483ffdd76755bc8dfc9f6/src/zmalloc.c#L380)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> zmalloc(size)
> ```
zmalloc_pmem
</blockquote></details>

Done.
___
*[src/zmalloc.c, line 401 at r7](https://reviewable.io/reviews/patkamin/redis/1#-MYjckOp6wEn4fimTqyQ:-MYobKQDCvsxiWL7Mbaj:b-896fix) ([raw file](https://github.com/patkamin/redis/blob/4fe40ce115b3595defc483ffdd76755bc8dfc9f6/src/zmalloc.c#L401)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
>     return (size < pmem_threshold) ? zmalloc_dram(size) : zmalloc_pmem(size);
> ```

I think this is a place for something like:

> ```
> #ifdef USE_MEMKIND_MEMTIER
>  return malloc(size);
> #else
>    return (size < pmem_threshold) ? zmalloc_dram(size) : zmalloc_pmem(size);
> #endif
> ```

But I am not sure what is the relation between pmem_threshold and MEMKIND_MEMTIER :/
</blockquote></details>

Done.
___
*[src/zmalloc.c, line 444 at r7](https://reviewable.io/reviews/patkamin/redis/1#-MYjd5ByDOMy2kHkYKbF:-MYomqnxAXX3o1CEvRSB:b-896fix) ([raw file](https://github.com/patkamin/redis/blob/4fe40ce115b3595defc483ffdd76755bc8dfc9f6/src/zmalloc.c#L444)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
>     return (size < pmem_threshold) ? zcalloc_dram(size) : zcalloc_pmem(size);
> ```
> #ifdef USE_MEMKIND_MEMTIER
>  return calloc(size);
> #else
>    return (size < pmem_threshold) ?  zcalloc_dram(size) : zcalloc_pmem(size);
> #endif
> ```
</blockquote></details>

Done.

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 2 unresolved discussions (waiting on @PatKamin)

___
*[src/zmalloc.c, line 249 at r8](https://reviewable.io/reviews/patkamin/redis/1#-MYszwBXEZwiMxgvrY1T:-MYszwBXEZwiMxgvrY1U:b-garwiy) ([raw file](https://github.com/patkamin/redis/blob/051fb28d6b649ce416d6273bebb7263432de06d9/src/zmalloc.c#L249)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> static int zmalloc_is_pmem(void * ptr) {
>     if (memory_variant == MEMORY_ONLY_DRAM) return DRAM_LOCATION;
>     struct memkind *temp_kind = memkind_detect_kind(ptr);
>     return (temp_kind == MEMKIND_DEFAULT) ? DRAM_LOCATION : PMEM_LOCATION;
> }
> ```
</details>

I think you could move zmalloc_is_pmem before zfree_pmem - inside #if defined(USE_MEMKIND) block - to be defined only in case of USE_MEMKIND - see my other comment
___
*[src/zmalloc.c, line 553 at r8](https://reviewable.io/reviews/patkamin/redis/1#-MYswtLaEZn8El6TWI-l:-MYswtLb-DLOCEAQdqbf:b-qdl9zh) ([raw file](https://github.com/patkamin/redis/blob/051fb28d6b649ce416d6273bebb7263432de06d9/src/zmalloc.c#L553)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
>     if (!zmalloc_is_pmem(ptr)) {
>         zfree_dram(ptr);
>     } else {
>         zfree_pmem(ptr);
>     }
> ```
</details>


This should be updated to something like:
<details><summary><i alt="summary">Quoted 9 lines of code…</i></summary>

> ```
> #ifdef USE_MEMKIND_MEMTIER
> memtier_free(ptr);
> #else
>     if (!zmalloc_is_pmem(ptr)) {
>         zfree_dram(ptr);
>     } else {
>         zfree_pmem(ptr);
>     }
>#endif
> ```
</details>


zmalloc_is_pmem can be even omitted in case of #ifdef USE_MEMKIND_MEMTIER

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 3 unresolved discussions (waiting on @PatKamin)

___
*[src/zmalloc.c, line 203 at r8](https://reviewable.io/reviews/patkamin/redis/1#-MYt-qg0Akg8w9kwjsBA:-MYt-qg0Akg8w9kwjsBB:b-kn5tvv) ([raw file](https://github.com/patkamin/redis/blob/051fb28d6b649ce416d6273bebb7263432de06d9/src/zmalloc.c#L203)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> #endif
> static int memory_variant = MEMORY_ONLY_DRAM;
> static size_t pmem_threshold = UINT_MAX;
> #ifndef USE_MEMKIND_MEMTIER
> ```
</details>

I think all of this could be inside one ifdef block

**Review by PatKamin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 3 unresolved discussions (waiting on @michalbiesek)

___
*[src/zmalloc.c, line 203 at r8](https://reviewable.io/reviews/patkamin/redis/1#-MYt-qg0Akg8w9kwjsBA:-MYt0haxBo8WBZuzT2OY:b-896fix) ([raw file](https://github.com/patkamin/redis/blob/051fb28d6b649ce416d6273bebb7263432de06d9/src/zmalloc.c#L203)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> #endif
> static int memory_variant = MEMORY_ONLY_DRAM;
> static size_t pmem_threshold = UINT_MAX;
> #ifndef USE_MEMKIND_MEMTIER
> ```
I think all of this could be inside one ifdef block
</blockquote></details>

Done.
___
*[src/zmalloc.c, line 249 at r8](https://reviewable.io/reviews/patkamin/redis/1#-MYszwBXEZwiMxgvrY1T:-MYt4FqlCAZ86UMafTZl:b-ouqyt4) ([raw file](https://github.com/patkamin/redis/blob/051fb28d6b649ce416d6273bebb7263432de06d9/src/zmalloc.c#L249)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> static int zmalloc_is_pmem(void * ptr) {
>     if (memory_variant == MEMORY_ONLY_DRAM) return DRAM_LOCATION;
>     struct memkind *temp_kind = memkind_detect_kind(ptr);
>     return (temp_kind == MEMKIND_DEFAULT) ? DRAM_LOCATION : PMEM_LOCATION;
> }
> ```
I think you could move zmalloc_is_pmem before zfree_pmem - inside #if defined(USE_MEMKIND) block - to be defined only in case of USE_MEMKIND - see my other comment
</blockquote></details>

Have to leave it this way in a commit which uses memtier API via kinds. Changed it in the next commit which uses memtier object though.
___
*[src/zmalloc.c, line 553 at r8](https://reviewable.io/reviews/patkamin/redis/1#-MYswtLaEZn8El6TWI-l:-MYt6gjl3u782GjpvTHK:b-896fix) ([raw file](https://github.com/patkamin/redis/blob/051fb28d6b649ce416d6273bebb7263432de06d9/src/zmalloc.c#L553)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
>     if (!zmalloc_is_pmem(ptr)) {
>         zfree_dram(ptr);
>     } else {
>         zfree_pmem(ptr);
>     }
> ```

This should be updated to something like:

> ```
> #ifdef USE_MEMKIND_MEMTIER
> memtier_free(ptr);
> #else
>     if (!zmalloc_is_pmem(ptr)) {
>         zfree_dram(ptr);
>     } else {
>         zfree_pmem(ptr);
>     }
>#endif
> ```

zmalloc_is_pmem can be even omitted in case of #ifdef USE_MEMKIND_MEMTIER

</blockquote></details>

Done.

**Review by michalbiesek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 3 unresolved discussions (waiting on @PatKamin)

___
*[src/zmalloc.c, line 210 at r9](https://reviewable.io/reviews/patkamin/redis/1#-MYxN7aJ9oNaxcy18gTO:-MYxN7aJ9oNaxcy18gTP:bvgzpbz) ([raw file](https://github.com/patkamin/redis/blob/9e9e67eda393ef6520e7788622f42cccf9c8f342/src/zmalloc.c#L210)):*
> ```
> static int memory_variant = MEMORY_ONLY_DRAM;
> static size_t pmem_threshold = UINT_MAX;
> ```
can this also be included inside the #ifndef USE_MEMKIND_MEMTIER above?
___
*[src/zmalloc.c, line 495 at r9](https://reviewable.io/reviews/patkamin/redis/1#-MYxMaFT666wHmI235sH:-MYxMaFT666wHmI235sI:bspllk4) ([raw file](https://github.com/patkamin/redis/blob/9e9e67eda393ef6520e7788622f42cccf9c8f342/src/zmalloc.c#L495)):*
> ```
> ····
> ```
nit remove whitespaces
___
*[src/zmalloc.c, line 499 at r9](https://reviewable.io/reviews/patkamin/redis/1#-MYxMZ6d2eFAPvm7seOt:-MYxMZ6d2eFAPvm7seOu:bspllk4) ([raw file](https://github.com/patkamin/redis/blob/9e9e67eda393ef6520e7788622f42cccf9c8f342/src/zmalloc.c#L499)):*
> ```
> ····
> ```
nit remove whitespaces

**Review by PatKamin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1)* status: 9 of 10 files reviewed, 1 unresolved discussion (waiting on @michalbiesek)

___
*[src/zmalloc.c, line 210 at r9](https://reviewable.io/reviews/patkamin/redis/1#-MYxN7aJ9oNaxcy18gTO:-MYxjsUR4PKPkN2yX8Xv:b4g03q2) ([raw file](https://github.com/patkamin/redis/blob/9e9e67eda393ef6520e7788622f42cccf9c8f342/src/zmalloc.c#L210)):*
<details><summary><i>Previously, michalbiesek wrote…</i></summary><blockquote>

> ```
> static int memory_variant = MEMORY_ONLY_DRAM;
> static size_t pmem_threshold = UINT_MAX;
> ```
can this also be included inside the #ifndef USE_MEMKIND_MEMTIER above?
</blockquote></details>

No, as these values are being called several times in server.c. IMHO it's not a big overhead to leave these values available in USE_MEMTIER scenario.

**Review by michalbiesek — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/patkamin/redis/1#-:-MYxpJZt8wuC5LnCAEY4:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @PatKamin)

**Pull request merged**